### PR TITLE
FEATURE: Introduced a flow node:status to monitor nodetree health

### PR DESCRIPTION
**Pull request by sbruggmann**
Corresponding to #1988

### CODE REVIEW

File: Neos.ContentRepository/Classes/Command/NodeCommandController.php [hunk 1]
```diff
@@ -118,6 +119,90 @@ public function repairCommand($nodeType = null, $workspace = 'live', $dryRun = f
...
+        $dryRun = true;
+        $summary = true;
+
+        $this->outputLine('Run status report..');
+
+        $this->pluginConfigurations = self::detectPlugins($this->objectManager);
+
+        if ($this->workspaceRepository->countByName($workspace) === 0) {
+            $this->outputLine('Workspace "%s" does not exist', [$workspace]);
+            exit(1);
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Command/NodeCommandController.php]:
```
$this->quit(1);
return;
```
would be the "proper" way - exit prevents the framework from being shut down correctly

File: Neos.ContentRepository/Classes/Command/NodeCommandController.php [hunk 2]
```diff
@@ -118,6 +119,90 @@ public function repairCommand($nodeType = null, $workspace = 'live', $dryRun = f
...
+        }
+
+        if ($nodeType !== null) {
+            if ($this->nodeTypeManager->hasNodeType($nodeType)) {
+                $nodeType = $this->nodeTypeManager->getNodeType($nodeType);
+            } else {
+                $this->outputLine('Node type "%s" does not exist', [$nodeType]);
+                exit(1);
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Command/NodeCommandController.php]: quit(1);

File: Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php [hunk 3]
```diff
@@ -291,25 +295,32 @@ protected function askBeforeExecutingTask($question, \Closure $task)
      *
      * @param string $workspaceName Name of the workspace to consider
      * @param boolean $dryRun Simulate?
+     * @param boolean $summary Only Status?
      * @param NodeType $nodeType Only for this node type, if specified
-     * @return void
+     * @return integer
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php]: What does the return value mean?

File: Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php [hunk 4]
```diff
@@ -291,25 +295,32 @@ protected function askBeforeExecutingTask($question, \Closure $task)
...
                     continue;
                 }
-                $this->createChildNodesByNodeType($nodeType, $workspaceName, $dryRun);
+                $errorCount += $this->createChildNodesByNodeType($nodeType, $workspaceName, $dryRun, $summary);
             }
         }
 
+        if ($summary) {
+            return $errorCount;
+        }
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php]: The `@return` annotation is `integer` but this only returns a value if `summary` is set?

**Comment by kitsunet** [comment]
Just as a short question, is this different enough to a node:repair --dry-run ? from the looks of it, it seems rather similar?

**Comment by bwaidelich** [comment]
FYI: I'll try to get a combination of this one and #2250 into a new PR using the `Result` return type as suggested.

**Comment by sbruggmann** [comment]
@kitsunet The big difference is, that the proposed output would be much cleaner and would allow to `grep` and script following actions.

**Comment by bwaidelich** [comment]
I looked into this, but it is really hard to implement this in a non breaking manner..
Using a `Result` object could work, but it has drawbacks:
* If we want to stay backwards compatible we still have to inject the `ConsoleOutput` and respect that implementations of `invokeSubCommand()` might not return a result
* Merging all results and outputting them would result in the repair command to only output anything after all tasks were executed

I think I found a better way:
We add a new interface `EventDispatchingNodeCommandControllerPluginInterface` (or some shorter name) and implement this in our two plugins.

Before:
```
$this->output->outputLine('Checking nodes of type "%s" for missing child nodes ...', [$nodeType->getName()]);
// ...
$this->output->outputLine('Node type "%s" does not exist', [(string)$nodeType]);
exit(1);
```
Afterwards:
```
$this->dispatch(self::EVENT_NOTICE, sprintf('Checking nodes of type "%s" for missing child nodes ...', $nodeType->getName()));
// ...
$this->dispatch(EventDispatcherInterface::EVENT_ERROR, sprintf('Node type "%s" does not exist', (string)$nodeType));
return;
```

That would be backwards compatible and extensible and for the next major we could finally cleanup the interfaces.

Also we should move from
```php
if ($dryRun === false) {
    $rootNode->createNode(NodePaths::getNodeNameFromPath(SiteService::SITES_ROOT_PATH));
    $this->output->outputLine('Missing "%s" node was created', [SiteService::SITES_ROOT_PATH]);
} else {
    $this->output->outputLine('"%s" node is missing!', [SiteService::SITES_ROOT_PATH]);
}
```
to sth like
```
$this->dispatch(self::EVENT_NOTICE, 'Create missing site node');
if ($dryRun === false) {
    $rootNode->createNode(NodePaths::getNodeNameFromPath(SiteService::SITES_ROOT_PATH));
    $this->dispatch(self::EVENT_SUCCESS);
}
```

What do you think?

**Comment by bwaidelich** [comment]
I gave it a try with #2278 (didn't want to mess with your PR)

**Comment by sbruggmann** [comment]
@bwaidelich thank you for taking care :-)

**Comment by bwaidelich** [comment]
@sbruggmann Thank you for starting it!
BTW: The current output of the reworked command is not _much_ cleaner, but with the foundation we can easily improve that (i.e. introduce a `verbose` flag etc).

**Pull request closed without merging**